core/src/wallet/algorand/database/AlgorandAccountDatabaseHelper.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * AlgorandAccountDatabaseHelper
+ *
+ * Created by Hakim Aammar on 18/05/2020.
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include "AlgorandAccountDatabaseHelper.hpp"
+#include "../model/AlgorandAccount.hpp"
+#include "../operations/AlgorandOperation.hpp"
+
+#include <wallet/common/AbstractAccount.hpp>
+#include <database/soci-number.h>
+
+namespace ledger {
+namespace core {
+namespace algorand {
+
+    std::string AccountDatabaseHelper::createAccount(soci::session & sql,
+                                                     const std::string & walletUid,
+                                                     const AccountDatabaseEntry & account) {
+
+        ledger::core::AccountDatabaseHelper::createAccount(sql, walletUid, account.index);
+
+        auto uid = createAccountUid(walletUid, account.index);
+
+        sql << "INSERT INTO algorand_accounts VALUES(:uid, :wallet_uid, :idx, :address)",
+            soci::use(uid),
+            soci::use(walletUid),
+            soci::use(account.index),
+            soci::use(account.address);
+
+        return uid;
+    }
+
+    bool AccountDatabaseHelper::queryAccount(soci::session & sql,
+                                             const std::string & accountUid,
+                                             AccountDatabaseEntry & account) {
+
+        soci::rowset<soci::row> rows = (sql.prepare << "SELECT idx, address "
+                                                       "FROM algorand_accounts "
+                                                       "WHERE uid = :uid", soci::use(accountUid));
+        for (auto &row : rows) {
+            account.index = soci::get_number<int32_t>(row, 0);
+            account.address = row.get<std::string>(1);
+        }
```

Comment:
Good point. While your solution would be perfectly fine, I chose to be consistent with the technique that's used everywhere in `XXXDatabaseHelpers`, by just adding a `return` inside the for loop.

Reply:
This is great to be consistent with our interface I couldn't agree more. However, for technical details just like the one mention here, feel free to improve it because there are some really old codes in our codebase and such piece of codes should disappear. By implementing new better code - bug-free and easier to read/understand - we introduce a *model* to fix the last helpers and ease the implementation of the new ones.

Reply:
I think you're right, however there's no easy way to check the size of a `soci::rowset` apparently. It's meant to be used only with iterators, which makes sense in a DB query context... and could explain the weird for loop used everywhere like this.

I think I can either use the for loop to dump all the results in a vector then check this vector, or maybe change the SQL code to retrieve only the 1st result and ignore any others... Does any of this suits you?